plugins/synapticsmst/synapticsmst-device.c

```diff
@@ -586,139 +1094,242 @@ synapticsmst_device_write_firmware (SynapticsMSTDevice *device,
+							"ESM firmware checksum error");
+			return FALSE;
+		}		
+	}
+	else{	
+		for (guint32 i = 0; i < (code_size + 17); i++)
+			checksum += *(payload_data + offset + i);
 
+		if (checksum & 0xFF) {
+			g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+						"firmware checksum error");
+			return FALSE;
+		}
+	}
 	/* TODO: May need a way to override this to cover field
 	 * issues of invalid firmware flashed*/
 	/* check firmware and board ID again */
 	tmp = (*(payload_data + ADDR_CUSTOMER_ID) << 8) + *(payload_data + ADDR_BOARD_ID);
-	if (tmp != synapticsmst_device_get_board_id (device)) {
-		g_set_error_literal (error,
-				     G_IO_ERROR,
-				     G_IO_ERROR_INVALID_DATA,
-				     "board ID mismatch");
-		return FALSE;
+	if(isChipPanamera == TRUE){
+		g_debug("bypass check CusID %x %x",tmp, synapticsmst_device_get_board_id (device));	
 	}
-
+	else {
+		if (tmp != synapticsmst_device_get_board_id (device)) {
+			g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+						"board ID mismatch");
+			return FALSE;
+		}
+	}
+	/* open device */
 	if (!synapticsmst_device_open (device, error)) {
 		g_prefix_error (error,
 				"can't open DP Aux node %s",
 				synapticsmst_device_get_aux_node (device));
 		return FALSE;
 	}
-
 	/* enable remote control */
 	if (!synapticsmst_device_enable_remote_control (device, error))
 		return FALSE;
 
-	/* erase SPI flash */
-	connection = synapticsmst_common_new (priv->fd, priv->layer, priv->rad);
-	if (synapticsmst_common_rc_set_command (connection,
-						UPDC_FLASH_ERASE,
-						2, 0, (guint8 *)&erase_code)) {
-		g_set_error_literal (error,
-				     G_IO_ERROR,
-				     G_IO_ERROR_INVALID_DATA,
-				     "can't erase flash");
-		return FALSE;
+	/* detect SPI quad mode & disable HDCP 2.2 */
+	if(isChipPanamera == TRUE)
+	{
+		/* Need to detect flash mode and ESM first ? */
+		/* disable flash Quad mode and ESM/HDCP2.2*/
+		for(guint8 i=0; i<16; i++){
+			dwData[i] = 0;	
+		}
+		rc = 0;
+		connection = synapticsmst_common_new (priv->fd, priv->layer, priv->rad);		
+		/* disable ESM first */
+		dwData[0] = 0x21;
+		rc = synapticsmst_common_rc_set_command (connection, UPDC_WRITE_TO_MEMORY,
+								4, (gint)0x2000fc, (guint8*)dwData);
+		g_usleep(200);// waiting for ESM exit
+		/* disable QUAD mode */ 
+		rc = synapticsmst_common_rc_get_command (connection, UPDC_READ_FROM_MEMORY,
+								((sizeof(dwData)/sizeof(dwData[0]))*4), (gint)0x200fc0, (guint8*)dwData);		
+		dwData[0] = 0x00;
+		rc = synapticsmst_common_rc_set_command (connection, UPDC_WRITE_TO_MEMORY,
+								4, (gint)0x200fc0, (guint8*)dwData);	
+		/* disable HDCP2.2 */
+		rc = synapticsmst_common_rc_get_command (connection, UPDC_READ_FROM_MEMORY,
+								4, (gint)0x200f90, (guint8*)dwData);		
+		dwData[ 0 ] = dwData[0] & (~BIT(2));
+		rc = synapticsmst_common_rc_set_command (connection, UPDC_WRITE_TO_MEMORY,
+								4, (gint)0x200f90, (guint8*)dwData);		
```

Comment:
Since these are magic offsets used many times, it would be good to add `#define` for them either in the header (if they'll be used in another file) or top of this source file.

Reply:
They are address of register, don't want to reveal with specific name. Could you please suggest a better to replace?

Reply:
I think just a `#define` with the purpose used in this plugin is fine (even if it supports other undocumented purpose)

For example `#define HDCP_CONTROL_REG 0x200f90`